### ISSUE TITLE
RFC - improved E2E pipeline

### ISSUE THREAD

**Issue by splincode** [Tinkoff/taiga-ui#1286]
### Discussed in https://github.com/Tinkoff/taiga-ui/discussions/864

<div type='discussions-op-text'>

<sup>Originally posted by **splincode** October 13, 2021</sup>
# Current behavior

Let's take a look at the current process. Let's say we have some initial state `hash#9eabf5b` of the main branch.

![image](https://user-images.githubusercontent.com/12021443/137077285-344db3ce-b708-4f3c-80a8-d97677dd088e.png)

- **Step 0**: At some point in time, developer `Any` pushes the code into the main branch. The Github Actions CI builds and deploys next demo app to the Github pages server on `taiga-ui.dev/next`;
- **Step 1**: Developer A creates the new branch `feat/ISSUE-123`;
- **Step 2**: Developer B creates the new branch `fix/ISSUE-456`;
- **Step 3**: Developer A makes some changes to his branch and pushes those changes: 
    - Then in order to make the e2e screenshot tests work it is necessary to prepare the original screenshots based on the main branch from `taiga-ui.dev/next`. This process takes 15 minutes (today);
    - After original screenshots are prepared, e2e tests run on the current branch codebase . This process also takes 15 minutes.
- **Step 4**: The developer `B` makes some changes to his branch and pushes those changes;
- **Step 5**: When the developer `B` pushes his code, the previous developer's branch is merged into the main branch which new state is `hash#29932f3`;
   - So, we know that for the developer `B` e2e tests were running before. However, the original screenshots were prepared based on the old state of the main branch (`hash#9eabf5b` while the new state `hash#29932f3`) from `taiga-ui.dev/next`. 
   -  In total, this again takes 30 minutes.
 - **Step 6**:  Only after the developer `B` makes new changes to the code and pushes his changes, the new version of `taiga-ui.dev/next` will show the correct state based on the main branch. Then we are sure that our e2e tests passed based on the new state;
 - **Step 7**: The branch of the developer `B` is merged into the main branch.

### The problems:

- The CI pipeline needs to generate original screenshots of the `taiga-ui.dev/next` in each pull request. In the future when more tests will be there, running tests will take twice more time.

**2021**: Prepare originals (15min) + compare with originals (15min);
**2022/2023**: Prepare originals (30min) + compare with originals (30min).

- Passing e2e tests on a developer branch does not guarantee that his tests were based on the actual state of the main branch.

- The e2e tests can be unstable in the developer's branch if the application base site URL is changed or there is a new deployment of application to the stand.

### The goals:

- Speed up the e2e pipeline on CI;
- Ensure consistency between branches;
- Simplify running scripts for e2e in CI.

# New behavior

The main idea is reuse prepared screenshots from the main branch.

![image](https://user-images.githubusercontent.com/12021443/137078375-5f4369c0-2d4f-4fc0-986b-25926941e7fa.png)

- **Step 0**: At some point in time, the developer `Any` pushes the code into the main branch. The Github Actions CI collected screenshots are sent to the artifacts server (our cache);
- **Step 1**: The developer `A` creates the new branch `feat/ISSUE-123`;
- **Step 2**: The developer `B` creates the new branch `fix/ISSUE-456`;
- **Step 3**: The developer `A` makes some changes to his branch and pushes those changes: 
    - The Github Actions CI simply pulls artifacts from the Github artifacts server (our cache);
    - Then e2e tests run on the current code base of the branch;
    - This process takes only 15 minutes instead of 30 minutes.
- **Step 4**: The developer B makes some changes to his branch and pushes those changes:
    - Same process as step 3 (only 15 minutes).
- **Step 5**: Developer's `A` branch was merged into the main branch; 
- **Step 6**: The developer `B` makes new changes to the code and pushes his changes;
    - Same process as step 3 (only 15 minutes) with new artifacts by hash@29932f3.
- **Step 7**: Developer B was merged into the main branch. 

### Pros:

- Halving the e2e tests run time;
- We run the e2e tests only with a locally running localhost:3333.

### Cons:

- When the main branch processes screenshot artifacts, we need to wait 15 minutes or less before new screenshots are uploaded to artifacts server. Only after the new state is uploaded, e2e pipeline runs to compare with new screenshots.

![image](https://user-images.githubusercontent.com/12021443/137079030-8a54695d-5e65-4ccf-abd8-81c364f40fde.png)

- Thus, after each merge to the main branch, we will have to wait for a new artifacts (our screenshot cache). It will take the same amount of time as before new behavior, about 30 minutes for each open pull requests. If you push some new changes to your branch and the main branch doesn't have new changes for rebase, remote artifacts are only fetched to current PR. The e2e tests will run for only 15 minutes.

# Next level of proposal

The main idea is to reuse prepared screenshots from the main branch, but only reuse artifacts that were collected in the previous branch before merging.

![image](https://user-images.githubusercontent.com/12021443/137079103-2146d103-1687-4921-bdef-5b096a006ade.png)

- **Step 0**: The cache is initially available;
- **Step 1**: The developer `A` creates the new branch `feat/ISSUE-123`;
- **Step 2**: The developer B creates the new branch `fix/ISSUE-456`;
- **Step 3**: The developer A makes some changes to his branch and pushes those changes: 
    - The Github Actions CI simply pulls artifacts from the Github artifacts server (our cache);
    - Then e2e tests run on the current code base of the branch;
    - This process takes only 15 minutes instead of 30 minutes;
    - A new cache is being formed by `hash@cff521` commit. 
- **Step 4**: The developer B makes some changes to his branch and pushes those changes:
    - The same process as step 3 (only 15 minutes);
    - A new cache is being formed by hash@1fe22f commit.
 - **Step 5**:  Merged into the main branch and now the default cache will be `hash@cff521`
 - **Step 6**:  Now there is no need to wait until screenshots are prepared in the main branch:
    - The cache `hash@cff521` that was from the previous merged branch will be used immediately;
    - The new cache is being formed by `hash@183aef` commit.
 - **Step 7**: Merged into the main branch and now the default cache will be `hash@183aef`.

### Pros:

- Halving the e2e tests run time;
- We run the e2e tests only with a locally running `localhost:3333`;
- No need to wait for the preparation of the cache in the main branch.

### Cons:

- It is necessary that developers' branches are always rebased onto the main branch for the cache to be up to date;
- It is necessary to implement additional checks if the cache will be lost in the main branch for some reason.

# What about parallelization for e2e tests?

This proposal is designed to reduce the time by revising the architecture for preparing the original screenshots.  At the next stage, it will be possible to consider the following development scenarios:

- Use Cypress Cloud
For the Github project we had free limit - 500 running tests in month.

![image](https://user-images.githubusercontent.com/12021443/137079692-af35ce8c-f192-4ac6-97c5-77775b16c98b.png)

However, we can use open source Cypress Cloud server - https://docs.sorry-cypress.dev/guide/get-started

- Additional help:
   - Upgrade to Cypress 8.5 and to Node.js 16, because in the new versions of V8  runtime performance has been improved;
   - Reduce up to 40% your Cypress suite execution time parallelizing the test run on the same machine - https://www.npmjs.com/package/cypress-parallel
   - Make Cypress Run Faster by Splitting Specs - https://glebbahmutov.com/blog/split-spec/
   - Boost your Cypress end-to-end Tests - https://andreybleme.com/2020-03-27/boost-your-cypress-tests/
   - How We Improved Network Speed By 300% In Cypress 3.3.0 - https://www.cypress.io/blog/2019/05/22/how-we-improved-network-speed-by-300-in-cypress-3-3-0/ </div>

**Comment by splincode** [comment]
Not relevant
